File: MAP/lab4/src/models/fileHandling/CloseRFile.java
package models.fileHandling;

import exceptions.fileExceptions.FileException;
import models.PrgState;
import models.expression.IExpression;
import models.statement.Statement;
import utils.IDictionary;

import java.io.BufferedReader;
import java.io.IOException;


public class CloseRFile implements Statement {

    private IExpression exp;

    public CloseRFile(IExpression exp){
        this.exp = exp;
    }

    @Override
    public PrgState execute(PrgState state) {

        IDictionary<String, Integer> dict = state.getDict();
        IFileTable<Integer, FileData> fileTable = state.getFileTable();
        int id = exp.eval(dict);


        if(!fileTable.contains(id))
            throw new FileException("Can't close the file because don't exist in FileTable");

        FileData fileData = fileTable.get(id);
        BufferedReader bufferedReader = fileData.getReader();

        try {
            bufferedReader.close();
        } catch (IOException e) {
            throw new FileException("Can't close the file");
        }

        fileTable.delete(id);
        //dict.delete();

        return state;
    }

    @Override
    public String toString(){
        return "close("+exp+")";
    }


}

File: MAP/lab7/lab7/Models/FileHandlers/CloseRFile.cs
using System;

namespace lab7.Models.FileHandlers
{
    public class CloseRFile : IStatement
    {
        private IExpression exp;

        public CloseRFile(IExpression exp)
        {
            this.exp = exp;
        }

        public PrgState execute(PrgState state)
        {
            int id = exp.eval(state.Dict);

            if (!state.FileTable.exists(id)) 
                throw new SystemException("This file" + exp + "can't be closed because do not exist");
            
            state.FileTable.get(id).Sw.Close();
            state.FileTable.delete(id);

            return state;
        }

        public override string ToString()
        {
            return "Close(" + exp + ")";
        }
    }
}